drivers/include/sx127x.h

```diff
@@ -0,0 +1,704 @@
+    uint32_t window_timeout;           /**< Timeout window */
+} sx127x_radio_settings_t;
+
+/**
+ * @brief   SX127X internal data.
+ */
+typedef struct {
+    /* Data that will be passed to events handler in application */
+    uint32_t last_channel;             /**< Last channel in frequency hopping sequence */
+    bool is_last_cad_success;          /**< Sign of success of last CAD operation (activity detected) */
+    xtimer_t tx_timeout_timer;         /**< TX operation timeout timer */
+    xtimer_t rx_timeout_timer;         /**< RX operation timeout timer */
+} sx127x_internal_t;
+
+/**
+ * @brief   SX127X hardware and global parameters.
+ */
+typedef struct {
+    spi_t spi;                         /**< SPI device */
+    gpio_t nss_pin;                    /**< SPI NSS pin */
+    gpio_t reset_pin;                  /**< Reset pin */
+    gpio_t dio0_pin;                   /**< Interrupt line DIO0 (Tx done) */
+    gpio_t dio1_pin;                   /**< Interrupt line DIO1 (Rx timeout) */
+    gpio_t dio2_pin;                   /**< Interrupt line DIO2 (FHSS channel change) */
+    gpio_t dio3_pin;                   /**< Interrupt line DIO3 (CAD done) */
+    gpio_t dio4_pin;                   /**< Interrupt line DIO4 (not used) */
+    gpio_t dio5_pin;                   /**< Interrupt line DIO5 (not used) */
+} sx127x_params_t;
+
+/**
+ * @brief   SX127X IRQ flags.
+ */
+typedef uint8_t sx127x_flags_t;
+
+/**
+ * @brief   SX127X device.
+ * @extends netdev_t
+ */
+typedef struct sx127x_s {
+    netdev_t netdev;                   /**< Netdev parent struct */
+    sx127x_radio_settings_t settings;  /**< Radio settings */
+    sx127x_params_t params;            /**< Device driver parameters */
+    sx127x_flags_t irq;                /**< Device IRQ flags */
+    sx127x_internal_t _internal;       /**< Internal sx127x data used within the driver */
+} sx127x_t;
+
+/**
+ * @brief   Hardware IO IRQ callback function definition.
+ */
+typedef void (sx127x_dio_irq_handler_t)(sx127x_t *dev);
+
+/**
+ * @brief   Setup the SX127X
+ *
+ * @param[in] dev                      Device descriptor
+ * @param[in] params                   Parameters for device initialization
+ */
+void sx127x_setup(sx127x_t *dev, const sx127x_params_t *params);
+
+/**
+ * @brief   Resets the SX127X
+ *
+ * @param[in] dev                      The sx127x device descriptor
+ */
+void sx127x_reset(const sx127x_t *dev);
+
+/**
+ * @brief   Initializes the transceiver.
+ *
+ * @param[in] dev                      The sx127x device descriptor
+ *
+ * @return result of initialization
+ */
+int sx127x_init(sx127x_t *dev);
+
+/**
+ * @brief   Initialize radio settings with default values
+ *
+ * @param[in] dev                      The sx127x device pointer
+ */
+void sx127x_init_radio_settings(sx127x_t *dev);
+
+/**
+ * @brief   Generates 32 bits random value based on the RSSI readings
+ *
+ * @pre     This function sets the radio in LoRa mode and disables all
```

Comment:
This is not a *pre*-condition, since it describes steps to be made after the function call. It's not even a postcondition (would be signified with `@post`), which would describe the state after execution. Best use `@attention` here.